Adjust required permissions for system schema

jon-wei opened the pull request:
This PR implements the following change described in #7563:

> The permissions needed for system table access could be adjusted to not require additional permissions beyond datasource-specific permissions for non-server related info (segments, tasks) and to require STATE-read permissions for server-related tables (servers, server_segments) to be consistent with the non-SQL APIs.

A STATE.READ permission check is added to `server_segments` sys table handling (`servers` already had one, and #7567 removed the unnecessary `sys` datasource check).

This PR also adds integration tests for sys schema + auth. A new segment and task for `auth_test` datasource are inserted directly into metadata during IT setup, and this datasource is used for testing the auth checks.

jon-wei commented:
will need to adjust tests to handle when the full IT test suite is run (since cluster state is not isolated across tests, things like "total stored segment size on historicals" change)

jon-wei commented on sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:
```diff
@@ -493,11 +493,30 @@ public TableType getJdbcTableType()
+          authorizerMapper
+      );
+      if (!stateAccess.isAllowed()) {
+        throw new ForbiddenException("Insufficient permission to view servers :" + stateAccess);
```
Fixed

jon-wei commented on sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:
```diff
@@ -493,11 +493,30 @@ public TableType getJdbcTableType()
+
+      final Access stateAccess = AuthorizationUtils.authorizeAllResourceActions(
+          authenticationResult,
+          Collections.singletonList(new ResourceAction(new Resource("STATE", ResourceType.STATE), Action.READ)),
```
Made this a constant in Resource

jon-wei commented on sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:
```diff
@@ -493,11 +493,30 @@ public TableType getJdbcTableType()
+          authenticationResult,
+          Collections.singletonList(new ResourceAction(new Resource("STATE", ResourceType.STATE), Action.READ)),
+          authorizerMapper
+      );
```
Moved this to a checkStateReadAccessForServers method

jon-wei commented:
I updated the SQL docs with info on necessary permissions, also added tests for a user without access to the `auth_test` datasource

gianm commented on docs/content/development/extensions-core/druid-basic-security.md:
```diff
@@ -310,6 +310,18 @@ For information on what HTTP methods are supported on a particular request endpo
+
+Queries on Druid datasources require DATASOURCE READ permissions for the specified datasource.
+
+Queries on the [information schema tables](../../querying/sql.html#information-schema) require DATASOURCE READ access for the specified datasource.
```
Please adjust the language a bit- there is not necessarily any 'specified datasource' with the information schema. The query might just be `SELECT * FROM INFORMATION_SCHEMA.COLUMNS`. I'd suggest:

```
Queries on the [INFORMATION_SCHEMA tables](../../querying/sql.html#information-schema) will
return information about datasources that the caller has DATASOURCE READ access to. Other
datasources will be omitted.
```

jon-wei replied:
Thanks, reworded

gianm commented:
LGTM other than the comment about doc wording.

jon-wei commented:
Fixed tests:
- Updated to handle changes from is_overshadowed patch
- Changed the created year for the inserted task to 2030 so it doesn't fall out of the completed tasks window
